sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/WebPubSubAsyncClient.java

```diff
@@ -0,0 +1,952 @@
+            .setAckId(ackId)
+            .setNoEcho(options.getNoEcho());
+
+        Mono<Void> sendMessageMono = sendMessage(message);
+        Mono<WebPubSubResult> responseMono = options.getFireAndForget()
+            ? sendMessageMono.then(Mono.just(new WebPubSubResult(null)))
+            : sendMessageMono.then(waitForAckMessage(ackId));
+        return responseMono.retryWhen(sendMessageRetrySpec);
+    }
+
+    /**
+     * Sends event.
+     *
+     * @param eventName the event name.
+     * @param content the data.
+     * @param dataType the data type.
+     * @return the result.
+     */
+    public Mono<WebPubSubResult> sendEvent(String eventName, BinaryData content, WebPubSubDataType dataType) {
+        return sendEvent(eventName, content, dataType, new SendEventOptions().setAckId(nextAckId()));
+    }
+
+    /**
+     * Sends event.
+     *
+     * @param eventName the event name.
+     * @param content the data.
+     * @param dataType the data type.
+     * @param options the options.
+     * @return the result.
+     */
+    public Mono<WebPubSubResult> sendEvent(String eventName, BinaryData content, WebPubSubDataType dataType,
+                                           SendEventOptions options) {
+        Objects.requireNonNull(eventName);
+        Objects.requireNonNull(content);
+        Objects.requireNonNull(dataType);
+        Objects.requireNonNull(options);
+
+        long ackId = options.getAckId() != null ? options.getAckId() : nextAckId();
+
+        BinaryData data = content;
+        if (dataType == WebPubSubDataType.BINARY || dataType == WebPubSubDataType.PROTOBUF) {
+            data = BinaryData.fromBytes(Base64.getEncoder().encode(content.toBytes()));
+        }
+
+        SendEventMessage message = new SendEventMessage()
+            .setEvent(eventName)
+            .setData(data)
+            .setDataType(dataType.name().toLowerCase(Locale.ROOT))
+            .setAckId(ackId);
+
+        Mono<Void> sendMessageMono = sendMessage(message);
+        Mono<WebPubSubResult> responseMono = options.getFireAndForget()
+            ? sendMessageMono.then(Mono.just(new WebPubSubResult(null)))
+            : sendMessageMono.then(waitForAckMessage(ackId));
+        return responseMono.retryWhen(sendMessageRetrySpec);
+    }
+
+    /**
+     * Receives group message events.
+     *
+     * @return the Publisher of group message events.
+     */
+    public Flux<GroupMessageEvent> receiveGroupMessageEvents() {
+        return groupMessageEventSink.asFlux();
+    }
+
+    /**
+     * Receives server message events.
+     *
+     * @return the Publisher of server message events.
+     */
+    public Flux<ServerMessageEvent> receiveServerMessageEvents() {
+        return serverMessageEventSink.asFlux();
+    }
+
+    /**
+     * Receives connected events.
+     *
+     * @return the Publisher of connected events.
+     */
+    public Flux<ConnectedEvent> receiveConnectedEvents() {
```

Comment:
I was wondering if `receiveConnectedEvents()`, `receiveDisconnectedEvents()`, `receiveStoppedEvents()` can merged into one sink, like a call `receiveServerActionEvents()`, since there may not be many events for each call and user needs to create three subcriber for each event?

Reply:
Actually only connected (and perhaps disconnected) is server action.

stopped is purely client (as reconnect/recover is disconnected but not stopped)

Reply:
Oh, ok,  my point is whether we can use one service call `receiveActionEvents()` to listen connected event, disconnected event and stopped event? so user can use one subscriber to handle all these events.

Reply:
Yeah, but what's the difference? Do switch case in the `receiveActionEvents()` seems not very user-friendly.

For example, I can think of user only listen to stopped, in case client not able to reconnect and need to start manually. And in this case they do not want to even care about connected, disconnected; then why bother about a combined receive?

Reply:
I was considering using a unified interface for them to receive all action events. But you are right, the user needs to write swithc cases in the subscriber, which seems not friendly for them. If the user has only one event to listen to, they can add a filter, but yes, that adds effort for the user. 

After thinking about it again, I am good with the current design.

Reply:
Yeah, the problem is that they are not actually "action" as well.

I would maybe group connected + disconnected (as they are from server, though disconnected might not...)

stopped is partly error handling (it called when user call stop, or reconnect failed)

There be a pending RejoinGroupFailedEvent which mostly an error handling.

So we probably want to revisit it later (it likely be premature to merge them now).